### DISCUSSION TITLE
About my link prediction project without true link labels

### DISCUSSION THREAD

**Discussion by ecastillot** [danielegrattarola/spektral#393] [Q&A]
Hi Spektral community

I want to do the next link prediction project:

I designed a graph where each node has 4 features and each one is connected each other with different weights. These weights was calculated by a formule invented by me in order to give more importance to some neighbors than others.  I want to compute node embeddings with a GCN, then applies a binary operator to those node embeddings to construct link embeddings. Those link embeddings are passed through a dense link classification layer to obtain link predictions between the same nodes.  

The problem is that I don't know how to fit the model. Normally, I have seen that I can fit the model with connections given in the network, for instance minimizing the binary cross-entropy between predicted link probabilities and true link label. However, I don't have true link labels. the link in my graph are weighted edges in order to give more importance to some neighbors than others when the node embeddings are being calculated. 

If you don't understand me, It will be enough if you answer this question for me:
I have a graph without edges, how can I design an unsupervised GNN to get link predictions between them?

Thanks a lot for your attention.

**Comment by danielegrattarola** [comment]
Link prediction can be seen as a similarity prediction, so if you want to go unsupervised you can look at some contrastive learning approaches for graphs. 

The better-known approach is probably GraphSAGE (https://github.com/williamleif/GraphSAGE) and, if you don't mind me plugging my own work, I've recently collaborated on a paper that does the same (https://arxiv.org/abs/2203.10866 -- in a revised version to be released soon we also do link prediction). 
In general, any approach to get unsupervised embeddings can be used for link prediction. 

Also from your description it seems that you could, for example, estimate the existence of edges by thresholding the edge weights. Would that work?
